MediaPlayer/MediaPlayer.js

```diff
@@ -391,16 +489,133 @@ const MediaPlayerBehaviorDecorator = hoc((config, Wrapped) => { // eslint-disabl
 		 * @memberof agate/MediaPlayer.MediaPlayerBase.prototype
 		 * @public
 		 */
-		loopChange = () => {
+		handleOnRepeat = () => {
+			// Handling the 3 states of repeat: repeat none, repeat one and repeat all
+			switch (this.state.repeat) {
+				case 'none':
+					this.setState(prevState  => {
+						return ({loop: !prevState.loop});
+					}, () => {
+						this.media.loop = this.state.loop;
+					});
+					this.setState({repeat: 'one'});
+					break;
+				case 'one':
+					this.setState({repeat: 'all'});
+					this.media.loop = false;
+					break;
+				case 'all':
+					this.setState({repeat: 'none'});
+					this.media.loop = false;
+			}
+		};
+
+		handleOnEnded = () => {
+			// Play next media only if current media is not last or, if it is last, only if repeat='all'
+			if (this.state.sourceIndex !== this.state.playlist.length - 1 || (this.state.sourceIndex === this.state.playlist.length - 1 && this.state.repeat === 'all')) {
+				this.handleNext();
+			}
+		};
+
+		handleNext = () => {
+			let currentIndex = this.state.sourceIndex;
+
+			if (currentIndex < this.state.playlist.length - 1) {
+				++currentIndex;
+			} else {
+				// When shuffle is true, the playback of the list restarts and the media list is reshuffled.
+				currentIndex = 0;
+
+				if (this.state.shuffle) {
+					this.shufflePlaylist();
+				}
+			}
+
+			this.setState(() => {
+				return ({sourceIndex: currentIndex});
+			}, () => {
+				this.play();
+			});
+		};
+
+		handlePrevious = () => {
+			let currentIndex = this.state.sourceIndex;
+
+			if (currentIndex > 0) {
+				--currentIndex;
+			} else if (!this.state.shuffle) {
+				currentIndex = this.state.playlist.length - 1;
+			}
+
+			this.setState(() => {
+				return ({sourceIndex: currentIndex});
+			}, () => {
+				this.play();
+			});
+		};
+
+		shufflePlaylist = (currentMedia) => {
+			let remainingSize = this.props.children.length;
+			let playlist = [...this.props.children];
+			let currentMediaIndex;
+
+			// While there are elements in the array
+			while (remainingSize > 0) {
+				// Pick a random index
+				let randomIndex = Math.floor(Math.random() * remainingSize);
+
+				// Decrease size by 1
+				remainingSize--;
+
+				// And swap the last element with it
+				let temp = playlist[remainingSize];
+				playlist[remainingSize] = playlist[randomIndex];
+				playlist[randomIndex] = temp;
+
+				if (playlist[remainingSize] === currentMedia) {
+					currentMediaIndex = remainingSize;
+				}
+			}
+
+			// Keep the current media active and set it as the first element in the shuffled array
+			if (currentMediaIndex) {
+				let temp = playlist[0];
+				playlist[0] = playlist[currentMediaIndex];
+				playlist[currentMediaIndex] = temp;
+			}
+
+			this.setState({
+				playlist,
+				sourceIndex: 0
+			});
+		};
+
+		handleShuffle = () => {
+			let currentMedia = this.state.playlist[this.state.sourceIndex];
+
```

Comment:
How about replacing `setState` to below?

```suggestion
			this.setState(({shuffle}) => {
				if (!shuffle) {
					return ({shuffle: true});
				} else {
					// When resetting shuffle to false, the initial playlist is set with the last played media kept active.
					return ({shuffle: false, playlist: this.props.children, sourceIndex: parseInt(currentMedia.key)});
				}
			}, () => {
				if (this.state.shuffle) {
					this.shufflePlaylist(currentMedia);
				}
			});
```